flocker/node/agents/testtools/_blockdevice.py

```diff
@@ -0,0 +1,871 @@
+        mountpoint directory.
+    """
+    check_output(['umount', unmount_target.path])
+
+
+def umount_all(root_path):
+    """
+    Unmount all devices with mount points contained in ``root_path``.
+
+    :param FilePath root_path: A directory in which to search for mount points.
+    """
+    def is_under_root(path):
+        try:
+            FilePath(path).segmentsFrom(root_path)
+        except ValueError:
+            return False
+        return True
+
+    partitions_under_root = list(p for p in psutil.disk_partitions()
+                                 if is_under_root(p.mountpoint))
+    for partition in partitions_under_root:
+        umount(FilePath(partition.mountpoint))
+
+
+def mountroot_for_test(test_case):
+    """
+    Create a mountpoint root directory and unmount any filesystems with mount
+    points beneath that directory when the test exits.
+
+    :param TestCase test_case: The ``TestCase`` which is being run.
+    :returns: A ``FilePath`` for the newly created mount root.
+    """
+    mountroot = FilePath(test_case.mktemp())
+    mountroot.makedirs()
+    test_case.addCleanup(umount_all, mountroot)
+    return mountroot
+
+
+def make_icloudapi_tests(
+        blockdevice_api_factory,
+):
+    """
+    :param blockdevice_api_factory: A factory which will be called
+        with the generated ``TestCase`` during the ``setUp`` for each
+        test and which should return a provider of both ``IBlockDeviceAPI``
+        and ``ICloudAPI`` to be tested.
+
+    :returns: A ``TestCase`` with tests that will be performed on the
+       supplied ``IBlockDeviceAPI``/``ICloudAPI`` provider.
+    """
+    class Tests(AsyncTestCase):
+        def setUp(self):
+            super(Tests, self).setUp()
+            self.api = blockdevice_api_factory(test_case=self)
+            self.this_node = self.api.compute_instance_id()
+            self.async_cloud_api = _SyncToThreadedAsyncCloudAPIAdapter(
+                _reactor=reactor, _sync=self.api,
+                _threadpool=reactor.getThreadPool())
+
+        def test_interface(self):
+            """
+            The result of the factory provides ``ICloudAPI``.
+            """
+            self.assertTrue(verifyObject(ICloudAPI, self.api))
+
+        def test_current_machine_is_live(self):
+            """
+            The machine running the test is reported as alive.
+            """
+            d = self.async_cloud_api.list_live_nodes()
+            d.addCallback(lambda live:
+                          self.assertIn(self.api.compute_instance_id(), live))
+            return d
+
+        def test_list_live_nodes(self):
+            """
+            ``list_live_nodes`` returns an iterable of unicode values.
+            """
+            live_nodes = self.api.list_live_nodes()
+            self.assertThat(live_nodes, AllMatch(IsInstance(unicode)))
+
+    return Tests
+
+
+@implementer(ICloudAPI)
+class FakeCloudAPI(proxyForInterface(IBlockDeviceAPI)):
```

Comment:
I guess it does? It's a `BlockDeviceAPI` fake for use with test infrastructure... _blockdevice has test infrastructure for `BlockDeviceAPI`'s ... yea I buy it.